### ISSUE TITLE
which-key performance issues

### ISSUE THREAD

**Issue by zaixi** [NvChad/NvChad#1321]
**Describe the bug**
When open a file from its last left off position feature is turned on and there are many which-key custom keymap, the cursor jumps slowly to restore the last position

**To Reproduce**
1. Set open a file from its last left off position autocmd
2. Customize 200  which-key mappings
3. Open file
4. cursor jumps slowly to restore the last position

**Expected behavior**
Cursor jumps fast to restore the last position.

**Screenshots**

**Desktop (please complete the following information):**
 - Ubuntu20.04
 - gnome-Terminal
 - NVIM v0.8.0-dev

**Additional context**

**Comment by siduck** [comment]
@LeonHeidelbach watch out for whichkey issues :angry:

**Comment by zaixi** [comment]
> @siduck @zaixi this is NOT a which-key issue. This is caused by the changes introduced [here](https://github.com/NvChad/NvChad/commit/5ba309a0dfee26b882658034c7bbf5ce03ebcde3). We had an issue back in the day where a user asked why some key bindings would be available in a buffer where they were not applicable (i.e. LSP bindings). In response to that @siduck wanted to implemented the behavior that key maps would only be applied on a per buffer basis which is the only way to achieve this, BUT as I've told you over discord before IT IS VERY SLOW. Especially since the key maps are sourced and applied every time a new buffer is created. The delay is a side effect of all mappings having to be sourced and applied first before any other action, like moving the caret to it's last saved document position, can be executed. Unless we change the mapping behavior back to the way it was before, i.e. applying all mappings only once on "VimEnter" rather than on every "BufCreate", issues like this aren't really "issues" but rather a known side effect.

I don't think so, I did an experiment locally, calling wk.register({ [keybind] = mapping_info }, opts) multiple times would make moving the caret to it's last saved document position slower, merging the mappings into one big table , calling wk.register once will significantly speed up

**Comment by zaixi** [comment]
> > @siduck @zaixi this is NOT a which-key issue. This is caused by the changes introduced [here](https://github.com/NvChad/NvChad/commit/5ba309a0dfee26b882658034c7bbf5ce03ebcde3). We had an issue back in the day where a user asked why some key bindings would be available in a buffer where they were not applicable (i.e. LSP bindings). In response to that @siduck wanted to implemented the behavior that key maps would only be applied on a per buffer basis which is the only way to achieve this, BUT as I've told you over discord before IT IS VERY SLOW. Especially since the key maps are sourced and applied every time a new buffer is created. The delay is a side effect of all mappings having to be sourced and applied first before any other action, like moving the caret to it's last saved document position, can be executed. Unless we change the mapping behavior back to the way it was before, i.e. applying all mappings only once on "VimEnter" rather than on every "BufCreate", issues like this aren't really "issues" but rather a known side effect.
> 
> I don't think so, I did an experiment locally, calling wk.register({ [keybind] = mapping_info }, opts) multiple times would make moving the caret to it's last saved document position slower, merging the mappings into one big table , calling wk.register once will significantly speed up

`--- a/lua/core/utils.lua
+++ b/lua/core/utils.lua
@@ -80,10 +80,15 @@ M.load_mappings = function(mappings, mapping_opt)
    -- set mapping function with/without whichkey
    local map_func
    local whichkey_exists, wk = pcall(require, "which-key")
+   local map_user = {}
 
    if whichkey_exists then
       map_func = function(keybind, mapping_info, opts)
-         wk.register({ [keybind] = mapping_info }, opts)
+          if map_user[opts.mode] == nil then
+              map_user[opts.mode] = {}
+          end
+          map_user[opts.mode][keybind] = mapping_info
+         -- wk.register({ [keybind] = mapping_info }, opts)
       end
    else
       map_func = function(keybind, mapping_info, opts)
@@ -113,6 +118,11 @@ M.load_mappings = function(mappings, mapping_opt)
          end
       end
    end
+   if whichkey_exists then
+       for mode, map in pairs(map_user) do
+           wk.register(map, { mode = mode })
+       end
+   end
 end
 
 -- remove plugins defined in chadrc
`

**Comment by zaixi** [comment]
> > @siduck @zaixi this is NOT a which-key issue. This is caused by the changes introduced [here](https://github.com/NvChad/NvChad/commit/5ba309a0dfee26b882658034c7bbf5ce03ebcde3). We had an issue back in the day where a user asked why some key bindings would be available in a buffer where they were not applicable (i.e. LSP bindings). In response to that @siduck wanted to implemented the behavior that key maps would only be applied on a per buffer basis which is the only way to achieve this, BUT as I've told you over discord before IT IS VERY SLOW. Especially since the key maps are sourced and applied every time a new buffer is created. The delay is a side effect of all mappings having to be sourced and applied first before any other action, like moving the caret to it's last saved document position, can be executed. Unless we change the mapping behavior back to the way it was before, i.e. applying all mappings only once on "VimEnter" rather than on every "BufCreate", issues like this aren't really "issues" but rather a known side effect.
> 
> I don't think so, I did an experiment locally, calling wk.register({ [keybind] = mapping_info }, opts) multiple times would make moving the caret to it's last saved document position slower, merging the mappings into one big table , calling wk.register once will significantly speed up


My local experiment
```
--- a/lua/core/utils.lua
+++ b/lua/core/utils.lua
@@ -80,10 +80,15 @@ M.load_mappings = function(mappings, mapping_opt)
    -- set mapping function with/without whichkey
    local map_func
    local whichkey_exists, wk = pcall(require, "which-key")
+   local map_user = {}
 
    if whichkey_exists then
       map_func = function(keybind, mapping_info, opts)
-         wk.register({ [keybind] = mapping_info }, opts)
+          if map_user[opts.mode] == nil then
+              map_user[opts.mode] = {}
+          end
+          map_user[opts.mode][keybind] = mapping_info
+         -- wk.register({ [keybind] = mapping_info }, opts)
       end
    else
       map_func = function(keybind, mapping_info, opts)
@@ -113,6 +118,11 @@ M.load_mappings = function(mappings, mapping_opt)
          end
       end
    end
+   if whichkey_exists then
+       for mode, map in pairs(map_user) do
+           wk.register(map, { mode = mode })
+       end
+   end
 end
 
 -- remove plugins defined in chadrc

```

**Comment by siduck** [comment]
@zaixi wk.register is needed for each mapping as we allow users to set opts for each mapping
